src/include/migraphx/op/nonmaxsuppression.hpp

```diff
@@ -128,100 +127,95 @@ struct nonmaxsuppression
-                    selected_indices.push_back(cidx);
-                    selected_indices.push_back(next_top_score.second);
-                }
-                sorted_boxes.pop();
-            }
-        });
-
-        result.visit([&](auto out) {
-            std::copy(selected_indices.begin(), selected_indices.end(), out.begin());
+        float iou_threshold   = (args.size() > 3) ? (args.at(3).at<float>()) : 0.0f;
+        float score_threshold = (args.size() > 4) ? (args.at(4).at<float>()) : 0.0f;
+
+        result.visit([&](auto output) {
+            visit_all(args[0], args[1])([&](auto boxes, auto scores) {
+                std::fill(output.begin(), output.end(), 0);
+                const auto& lens       = scores.get_shape().lens();
+                const auto num_batches = lens[0];
+                const auto num_classes = lens[1];
+                const auto num_boxes   = boxes.get_shape().lens()[1];
+                // boxes of a class with NMS applied [score, index]
+                std::vector<std::pair<float, int64_t>> selected_boxes_inside_class;
+                std::vector<int64_t> selected_indices;
+                selected_boxes_inside_class.reserve(output_shape.elements());
+                // iterate over batches and classes
+                shape comp_s{shape::float_type, {num_batches, num_classes}};
+                shape_for_each(comp_s, [&](auto idx) {
+                    auto batch_idx = idx[0];
+                    auto class_idx = idx[1];
+                    // index offset for this class
+                    std::size_t score_offset_ind =
+                        (batch_idx * num_classes + class_idx) * num_boxes;
+                    // index to first value of this batch
+                    std::size_t batch_boxes_ind = batch_idx * num_boxes * 4;
+                    std::priority_queue<std::pair<float, int64_t>> boxes_heap;
+                    filter_boxes_by_score(
+                        scores, score_offset_ind, num_boxes, score_threshold, boxes_heap);
+                    selected_boxes_inside_class.clear();
+                    // Get the next box with top score, filter by iou_threshold
+                    while(!boxes_heap.empty() &&
+                          selected_boxes_inside_class.size() < max_output_boxes_per_class)
+                    {
+                        // Check with existing selected boxes for this class, remove box if it
+                        // exceeds the IOU (Intersection Over Union) threshold
+                        const auto next_top_score = boxes_heap.top();
+                        bool not_selected         = std::any_of(
+                            selected_boxes_inside_class.begin(),
+                            selected_boxes_inside_class.end(),
+                            [&](auto selected_index) {
+                                return this->suppress_by_iou(
+                                    batch_box(boxes, batch_boxes_ind, next_top_score.second),
```

Comment:
It would be better to pass in the start with `boxes.begin() + (batch_idx * num_boxes * 4)`.